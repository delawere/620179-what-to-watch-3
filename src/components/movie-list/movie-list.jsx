import React, {PureComponent} from 'react';
import {arrayOf, exact, string, func} from 'prop-types';
import MovieCard from '../movie-card/movie-card.jsx';

class MovieList extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeCard: null
    };

    this.handleCardMouseOver = this.handleCardMouseOver.bind(this);
  }

  handleCardMouseOver(cardName) {
    this.setState({
      activeCard: cardName
    });
  }

  render() {
    const {films, onClick} = this.props;

    return (
      <div className="catalog__movies-list">
        {films.map(({name, img}) => (
          <MovieCard key={name} name={name} img={img} onClick={onClick} onMouseOver={this.handleCardMouseOver}/>
        ))}
      </div>
    );
  }
}


MovieList.propTypes = {
  films: arrayOf(exact({
    name: string,
    img: string
  })),
  onClick: func,
  onMouseOver: func
};

export default MovieList;
