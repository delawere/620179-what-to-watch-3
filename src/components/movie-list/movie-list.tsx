// Libs
import * as React from 'react';
import {connect} from "react-redux";
// Utils
import {FilmsType, FilmType} from '../../types';
import {getShownCardsNumber} from "../../reducer/films/selectors.js";
// Components
import MovieCard from "../movie-card/movie-card";

interface Props {
  activeItem?: FilmType;
  filteredFilms: FilmsType;
  filter?: string;
  onOpenCard: () => void;
  shownCardsNumber: number;
  setActiveItem: () => void;
  removeActiveItem: () => void;
  setTimer: () => void;
  removeTimer: () => void;
  maxCardsNumber?: number; 
  openedFilmId?: number;
}

const MovieList = (props: Props) => {
  const {shownCardsNumber, activeItem, filteredFilms, maxCardsNumber, openedFilmId} = props;
  const shownFilms = filteredFilms.slice(0, 
    maxCardsNumber === undefined 
      ? shownCardsNumber 
      : maxCardsNumber
  );

  const getFilmsWithoutOpened = (films, openedId) => films.filter(({id}) => id !== parseInt(openedId, 10));

  return (
    <div className="catalog__movies-list">
      {getFilmsWithoutOpened(shownFilms, openedFilmId)
        .map((film) => {
        const {name, id} = film;
        return (
          <MovieCard
            key={id}
            active={name === activeItem.name}
            {...film}
            {...props}
          />
        );
      })}
    </div>
  );
};

const mapStateToProps = (state) => ({
  shownCardsNumber: getShownCardsNumber(state)
});

export {MovieList};
export default connect(mapStateToProps)(React.memo(MovieList));
