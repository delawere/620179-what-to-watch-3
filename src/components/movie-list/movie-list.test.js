import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import {MovieList} from './movie-list';
import {BrowserRouter as Router} from 'react-router-dom';
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

const films = [
  {
    name: `name1`,
    img: `img/name1.jpg`,
  },
  {
    name: `name2`,
    img: `img/name2.jpg`,
  },
  {
    name: `name3`,
    img: `img/name3.jpg`,
  },
];

it(`MovieList renders correctly`, () => {
  const store = mockStore({
    filmsByGenre: films,
  });

  const {filmsByGenre} = store.getState();

  const tree = renderer
    .create(
        <Router>
          <Provider store={store}>
            <MovieList filmsByGenre={filmsByGenre}/>
          </Provider>
        </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
