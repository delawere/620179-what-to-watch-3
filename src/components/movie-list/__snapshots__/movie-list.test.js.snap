// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList renders correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="name1"
        height="175"
        src="img/name1.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        name1
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="name2"
        height="175"
        src="img/name2.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        name2
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="name3"
        height="175"
        src="img/name3.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        name3
      </a>
    </h3>
  </article>
</div>
`;
